pkg/operator/ceph/cluster/osd/osd.go

```diff
@@ -325,13 +325,32 @@ func (c *Cluster) postReconcileUpdateOSDProperties(desiredOSDs map[int]*OSDInfo)
 		return errors.Wrap(err, "failed to get osd usage")
 	}
 	logger.Debugf("post processing osd properties with %d actual osds from ceph osd df and %d existing osds found during reconcile", len(osdUsage.OSDNodes), len(desiredOSDs))
+	atleastOneOsdGotResize := false
 	for _, actualOSD := range osdUsage.OSDNodes {
-		if desiredOSD, ok := desiredOSDs[actualOSD.ID]; ok {
-			if err := c.updateDeviceClassIfChanged(actualOSD.ID, desiredOSD.DeviceClass, actualOSD.DeviceClass); err != nil {
+		if !c.spec.Storage.AllowOsdCrushWeightUpdate {
+			value, err := cephclient.ResizeOsdCrushWeight(actualOSD, c.context, c.clusterInfo)
+			if err != nil {
 				// Log the error and allow other updates to continue
-				logger.Error(err)
+				logger.Errorf("failed to resize osd crush weight on cluster in namespace %s: %v", c.clusterInfo.Namespace, err)
+			}
+			if value {
+				atleastOneOsdGotResize = true
 			}
 		}
+
+		desiredOSD, ok := desiredOSDs[actualOSD.ID]
+		if !ok {
+			continue
+		}
+		if err := c.updateDeviceClassIfChanged(actualOSD.ID, desiredOSD.DeviceClass, actualOSD.DeviceClass); err != nil {
+			// Log the error and allow other updates to continue
+			logger.Errorf("failed to update device class on cluster in namespace %s: %v", c.clusterInfo.Namespace, err)
+		}
+	}
+
+	if err := cephclient.ResizeCrushTreeBucketWeight(atleastOneOsdGotResize, c.context, c.clusterInfo); err != nil {
```

Comment:
There is a potential race condition here where Rook reweights all the OSDs, but then the reconcile fails (or stops for any reason) before this method is called or completed. In that case, Rook may not try to run this command on the next reconcile. Can we detect whether reweight is necessary via some other means?

Reply:
I can move this check above the `updateDeviceClassIfChanged` so which makes it no error prone for race conditions, 
As if the OSD is resizing successfully this will be called,

But if the methods fail during this `rewight-all` cmd then we dont do anything just print error, and in future we may not look back to it.


Considering ways for it,

1) We may get the actaull subtree bucket weight, but for calculating the desired state would be little complicated(like running a BFS on tree to get the weight of the subree)

Reply:
Removed it as it is not required